packages/block-editor/src/store/effects.js

```diff
@@ -84,12 +84,31 @@ export default {
 		const blockB = getBlock( state, clientIdB );
 		const blockBType = getBlockType( blockB.name );
 		const { clientId, attributeKey, offset } = getSelectionStart( state );
-		const hasTextSelection = (
+		const selectedBlockType = clientId === clientIdA ? blockAType : blockBType;
+		const attributeDefinition = selectedBlockType.attributes[ attributeKey ];
+		const canRestoreTextSelection = (
 			( clientId === clientIdA || clientId === clientIdB ) &&
 			attributeKey !== undefined &&
-			offset !== undefined
+			offset !== undefined &&
+			// We cannot restore text selection if the RichText identifier
+			// is not a defined block attribute key. This can be the case if the
+			// fallback intance ID is used to store selection (and no RichText
+			// identifier is set), or when the identifier is wrong.
+			attributeDefinition
```

Comment:
Very minor nitpick, but I think it'd be preferable to coerce this to a boolean so that the type matches what the name of the variable implies.

```suggestion
			!! attributeDefinition
```